lisa/notifier.py

```diff
@@ -0,0 +1,85 @@
+from dataclasses import dataclass
+from typing import Any, Dict, List, Type
+
+from lisa import schema
+from lisa.util import InitializableMixin, subclasses
+from lisa.util.logger import get_logger
+
+
+@dataclass
+class MessageBase:
+    type: str = ""
+    elapsed: float = 0
+
+
+class Notifier(subclasses.BaseClassWithRunbook, InitializableMixin):
+    def __init__(self, runbook: schema.TypedSchema) -> None:
+        super().__init__(runbook=runbook)
+        self._log = get_logger("notifier", self.__class__.__name__)
+
+    def finalize(self) -> None:
+        """
+        All test done. notifier should release resource,
+        or do finalize work, like save to a file.
+
+        Even failed, this method will be called.
+        """
+        pass
+
+    def _subscribed_message_type(self) -> List[Type[MessageBase]]:
+        raise NotImplementedError("must specify supported message types")
+
+    def _received_message(self, message: MessageBase) -> None:
+        raise NotImplementedError
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        """
+        initialize is optional
+        """
+        pass
+
+
+_notifiers: List[Notifier] = []
+_messages: Dict[type, List[Notifier]] = dict()
+
+
+# below methods uses to operate a global notifiers,
+# so that any object can send messages.
+def initialize(runbooks: List[schema.Notifier]) -> None:
+
+    factory = subclasses.Factory[Notifier](Notifier)
+    log = get_logger("init", "notifier")
+    for runbook in runbooks:
+        notifier = factory.create_by_runbook(runbook=runbook)
+        _notifiers.append(notifier)
+
+        subscribed_message_types: List[
+            Type[MessageBase]
+        ] = notifier._subscribed_message_type()
+
+        for message_type in subscribed_message_types:
+            registered_notifiers = _messages.get(message_type, [])
+            registered_notifiers.append(notifier)
+            _messages[message_type] = registered_notifiers
+        log.debug(
+            f"registered [{notifier.type_name()}] "
+            f"on messages: [{[x.type for x in subscribed_message_types]}]"
+        )
+
+        notifier.initialize()
+
+
+def notify(message: MessageBase) -> None:
```

Comment:
The intent of this is to allow the runner to send notifications, which right now are just printed messages, but _may be_ stuff like emails, CI notifications, etc., right? I'm not super sure how I feel about making this a kitchen sink. CI Pipelines can and should handle the downstream notifications (like emails, if I'm actually understanding this correctly), since it's already implemented and maintained by that service. LISA should probably stop at printing messages and returning an exit code. @lizzha what do you think?

Reply:
The lisa is designed to be able to integrate in some pipeline like ado, or run standalone. Another requirement is to save test result to database like lisav2. Other users can also extend lisa to save result in their existing test result schema.